Library/Homebrew/formula_cellar_checks.rb

```diff
@@ -325,11 +325,34 @@ def check_binary_arches(formula)
     end
     return if mismatches.empty?
 
-    <<~EOS
-      Binaries built for a non-native architecture were installed into #{formula}'s prefix.
-      The offending files are:
-        #{mismatches * "\n        "}
-    EOS
+    compatible_universal_binaries = mismatches.select do |file|
+      file.arch == :universal && file.archs.include?(Hardware::CPU.arch)
+    end
+    mismatches -= compatible_universal_binaries
+
+    universal_binaries_expected =
+      formula.tap.present? && tap_audit_exception(:universal_binary_allowlist, formula.name)
+    return if mismatches.empty? && universal_binaries_expected
+
+    s = ""
+
+    unless mismatches.empty?
+      s += <<~EOS
+        Binaries built for an incompatible architecture were installed into #{formula}'s prefix.
+        The offending files are:
+          #{mismatches * "\n  "}
+      EOS
+    end
+
+    if !compatible_universal_binaries.empty? && !universal_binaries_expected
+      s += <<~EOS
+        Unexpected universal binaries were found.
+        The offending files are:
+          #{compatible_universal_binaries * "\n  "}
+      EOS
+    end
```

Comment:
Currently, I believe you'll see this error if you `brew json llvm` without Homebrew/core tapped. Not sure if that's what we want, but we can always adjust it, perhaps at a later date when we roll out being able to install formulae without Homebrew/core tapped.

CC @Rylan12

Reply:
I suppose we could write to the tab whether a formula expects universal binaries...

Reply:
Tab won't show up in JSON files we upload but will be downloaded/cached when installing the bottle.

Reply:
> I suppose we could write to the tab whether a formula expects universal binaries... Does that show up in the JSON files we upload to https://formulae.brew.sh?

Yeah, this is just the JSON version of `brew info`

> Currently, I believe you'll see this error if you `brew json llvm` without Homebrew/core tapped

Why is that?

I don't think this should be much of a long-term issue because `brew audit` is a developer command so anyone running it needs to have a full clone of homebrew/core and won't be using `HOMEBREW_JSON_CORE`. Eventually, there will be an error if a `HOMEBREW_DEVELOPER` or `devcmdrun` user sets `HOMEBREW_JSON_CORE` (but I figured for testing purposes that will be easier to add in the future).

Reply:
> Why is that?

Because I see it when I `brew install llvm` without the changes here. Is that just because I have `HOMEBREW_DEVELOPER` set?

Reply:
Oh, is it not a problem anymore in the latest commit?

Hmm, that I don't know. It shouldn't be a problem that you have `HOMEBREW_DEVELOPER` set but I guess I haven't done extensive testing of `HOMEBREW_JSON_CORE` with `HOMEBREW_DEVELOPER`. I don't really understand why it would be any different whether you use `HOMEBREW_JSON_CORE` or not because its installing the same bottle regardless. This error seems to be dependent on the files themselves, so I don't know why it would be any different.

If it's still an issue I can probably test it out, if not I'd say just ignore it.

Reply:
Yea, the extra checks seem to be coming from having `HOMEBREW_DEVELOPER` set. The audit isn't run at install time if I unset it.